File: Hackerrank/3_month_preparation/python/week5.py
def max_min(k, arr):
    # Time complexity: O(n*log(n))
    # Space complexity (ignoring input): O(1)
    arr.sort()
    minimum_unfairness = arr[k - 1] - arr[0]
    for index in range(1, len(arr) - k + 1):
        if minimum_unfairness > arr[index + k - 1] - arr[index]:
            minimum_unfairness = arr[index + k - 1] - arr[index]

    return minimum_unfairness


def strong_password(n, password: str):
    # Time complexity: O(n)
    # Space complexity (ignoring input): O(1)
    add_lower = True
    add_upper = True
    add_number = True
    add_special = True
    for letter in password:
        if letter.islower():
            add_lower = False
        if letter.isupper():
            add_upper = False
        if letter.isnumeric():
            add_number = False
        if letter in "!@#$%^&*()-+":
            add_special = False

    characters_to_add = 0
    if add_lower:
        characters_to_add += 1
    if add_upper:
        characters_to_add += 1
    if add_number:
        characters_to_add += 1
    if add_special:
        characters_to_add += 1

    if len(password) + characters_to_add < 6:
        return 6 - len(password)

    return characters_to_add


def dynamic_array(n, queries):
    # Time complexity: O(n+q)
    # Space complexity (ignoring input): O(n+q)
    array = []
    for _ in range(0, n):
        array.append([])
    last_answer = 0
    result = []
    for query in queries:
        if query[0] == 1:
            array[(query[1] ^ last_answer) % n].append(query[2])

        if query[0] == 2:
            idx = (query[1] ^ last_answer) % n
            last_answer = array[idx][query[2] % len(array[idx])]
            result.append(last_answer)

    return result


def is_smart_number(num):
    val = int(math.sqrt(num))
    # Just change the line below
    if num / val == val:
        return True
    return False


def missing_numbers(arr, brr):
    # Time complexity: O(a + b)
    # Space complexity (ignoring input): O(a + b)
    arr_dict = {}
    for value in arr:
        if value in arr_dict:
            arr_dict[value] += 1
        else:
            arr_dict[value] = 1

    brr_dict = {}
    for value in brr:
        if value in brr_dict:
            brr_dict[value] += 1
        else:
            brr_dict[value] = 1

    missing_values = []
    for key in brr_dict.keys():
        if key in arr_dict:
            if arr_dict[key] <= brr_dict[key]:
                missing_values.append(key)
        else:
            missing_values.append(key)

    missing_values.sort()
    return missing_values


def full_counting_sort(arr):
    # Time complexity: O(n)
    # Space complexity (ignoring input): O(1)
    sorted_array = [""] * 101
    for index in range(0, len(arr)):
        sorted_index = int(arr[index][0])
        if index < len(arr) / 2:
            sorted_array[sorted_index] += " -"
        else:
            sorted_array[sorted_index] += " "
            for letter in arr[index][1]:
                sorted_array[sorted_index] += letter

    result_string = ""
    for string in sorted_array:
        result_string += string

    print(result_string[1:])


def grid_challenge(grid):
    #Time complexity: O(n^2*log(n))
    #Space complexity (ignoring input): O(n)
    sorted_grid = []
    for string in grid:
        sorted_grid.append(sorted(string))

    for row in range(0, len(grid) - 1):
        for column in range(0, len(grid[0])):
            if sorted_grid[row][column] > sorted_grid[row + 1][column]:
                return "NO"

    return "YES"


def sansa_and_xor(arr):
    #Time complexity: O(n)
    #Space complexity (ignoring input): O(1)
    #A number will appear in {(index+1)*(n-index)} subsequences
    #In case arr.len() is par, the appearences of any number will be par as well
    xor_value = 0
    for index in range(0, len(arr)):
        frequency_number = (index + 1) * (len(arr) - index)
        if frequency_number % 2 != 0:
            xor_value ^= arr[index]

    return xor_value

File: Hackerrank/3_month_preparation/rust/src/week5.rs
fn max_min(k: i32, arr: &[i32]) -> i32 {
    //Time complexity: O(n*log(n))
    //Space complexity (ignoring input): O(n)
    let mut arr = arr.to_vec();
    let k = k as usize;
    arr.sort_unstable();
    let mut minimun_unfairness = arr[k - 1] - arr[0];
    for index in 1..(arr.len() - k + 1) {
        if minimun_unfairness > arr[k - 1 + index] - arr[index] {
            minimun_unfairness = arr[k - 1 + index] - arr[index];
        };
    }
    minimun_unfairness
}

fn strong_password(n: i32, password: &str) -> i32 {
    //Time complexity: O(n)
    //Space complexity (ignoring input): O(1)
    let special_characters: std::collections::HashSet<char> = "!@#$%^&*()-+".chars().collect();

    let mut add_lower = true;
    let mut add_upper = true;
    let mut add_special = true;
    let mut add_number = true;
    for letter in password.chars() {
        if letter.is_uppercase() {
            add_upper = false;
        }
        if letter.is_lowercase() {
            add_lower = false;
        }
        if letter.is_numeric() {
            add_number = false;
        }
        if special_characters.contains(&letter) {
            add_special = false;
        }
    }

    let mut characters_to_add: i32 = 0;

    if add_lower {
        characters_to_add += 1;
    }
    if add_upper {
        characters_to_add += 1;
    }
    if add_special {
        characters_to_add += 1;
    }
    if add_number {
        characters_to_add += 1;
    }

    if (password.len() + characters_to_add as usize) < 6 {
        return 6 - password.len() as i32;
    }
    characters_to_add
}

fn dynamic_array(n: i32, queries: &[Vec<i32>]) -> Vec<i32> {
    //Time complexity: O(n+q)
    //Space complexity (ignoring input): O(n+q)
    let mut array = Vec::with_capacity(n as usize);
    for _ in 0..n {
        array.push(Vec::new());
    }
    let mut result_array = Vec::new();
    let mut last_answer = 0;
    for query in queries {
        let idx = ((query[1] ^ last_answer) % n) as usize;
        if query[0] == 1 {
            array[idx].push(query[2])
        }
        if query[0] == 2 {
            let size = query[2] as usize % (array[idx].len());
            last_answer = array[idx][size];
            result_array.push(last_answer);
        }
    }
    result_array
}

fn missing_numbers(arr: &[i32], brr: &[i32]) -> Vec<i32> {
    //Time complexity: O(a+b)
    //Space complexity (ignoring input): O(a+b)
    let mut arr_hash = std::collections::HashMap::new();
    for value in arr {
        match arr_hash.get(value) {
            Some(f) => arr_hash.insert(value, f + 1),
            None => arr_hash.insert(value, 1),
        };
    }
    let mut brr_hash = std::collections::HashMap::new();
    for value in brr {
        match brr_hash.get(value) {
            Some(f) => brr_hash.insert(value, *f + 1),
            None => brr_hash.insert(value, 1),
        };
    }

    let mut missing_values = Vec::new();
    for (&key, frequency_b) in brr_hash.iter() {
        let frequency_a = arr_hash.get(key).unwrap_or(&0);
        if frequency_a <= frequency_b {
            missing_values.push(*key);
        }
    }

    missing_values.sort_unstable();
    missing_values
}

fn full_countint_sort(arr: &[Vec<String>]) {
    //Time complexity: O(n)
    //Space complexity (ignoring input): O(1)
    let mut sorted_array: Vec<String> = vec!["".to_string(); 101];
    for index in 0..arr.len() {
        let sorted_index = arr[index][0]
            .parse::<usize>()
            .expect("To be able to parse correctly");
        if index < arr.len() / 2 {
            sorted_array[sorted_index].push_str(" -");
        } else {
            sorted_array[sorted_index].push(' ');
            sorted_array[sorted_index].push_str(&arr[index][1]);
        }
    }
    let mut result_array = String::new();
    for string in sorted_array {
        if !string.is_empty() {
            if result_array.is_empty() {
                result_array = string[1..].to_string();
            } else {
                result_array.push_str(&string);
            }
        }
    }
    println!("{}", result_array);
}

fn grid_challenge(grid: &[String]) -> String {
    //Time complexity: O(n^2*log(n))
    //Space complexity (ignoring input): O(n)
    let mut sorted_grid = Vec::new();
    for string in grid {
        let mut vec_chars = string.chars().collect::<Vec<char>>();
        vec_chars.sort_unstable();
        sorted_grid.push(vec_chars);
    }
    for row in 0..sorted_grid.len() - 1 {
        for column in 0..sorted_grid[0].len() {
            if sorted_grid[row][column] > sorted_grid[row + 1][column] {
                return "NO".to_string();
            }
        }
    }
    return "YES".to_string();
}
fn sansa_and_xor(arr: &[i32]) -> i32 {
    //Time complexity: O(n)
    //Space complexity (ignoring input): O(1)
    //A number will appear in {(index+1)*(n-index)} subsequences
    //In case arr.len() is par, the appearences of any number will be par as well
    if arr.len() % 2 == 0 {
        return 0;
    }
    let mut xor_value = 0;
    for index in 0..arr.len() {
        let appearences = (index + 1) * (arr.len() - index);
        if appearences % 2 != 0 {
            xor_value ^= arr[index]
        }
    }
    return xor_value;
}
